include/revng/Pipes/InvalidationHelpers.h

```diff
@@ -0,0 +1,152 @@
+#include "revng/Pipeline/LLVMContainer.h"
+#include "revng/Pipeline/Location.h"
+#include "revng/Pipeline/Target.h"
+#include "revng/Pipes/FileContainer.h"
+#include "revng/Pipes/FunctionKind.h"
+#include "revng/Pipes/Kinds.h"
+#include "revng/Pipes/ModelGlobal.h"
+#include "revng/Pipes/TaggedFunctionKind.h"
+#include "revng/Support/FunctionTags.h"
+#include "revng/TupleTree/TupleTreeDiff.h"
+#include "revng/Yield/CrossRelations.h"
+
+using yield::CrossRelations;
+
+namespace revng::kinds {
+
+// Find callers of a given `model::Function` via CrossRelations
+inline void insertCallers(const TupleTree<CrossRelations> &CR,
+                          MetaAddress Entry,
+                          std::set<MetaAddress> &Result) {
+  const auto Location = location(revng::ranks::Function, Entry);
+  const auto &Related = CR->Relations.at(Location.toString()).Related;
+  for (const yield::RelationTarget &RT : Related) {
+    if (RT.Kind == yield::RelationType::IsCalledFrom) {
+      auto Caller = genericLocationFromString<0>(RT.Location,
+                                                 revng::ranks::Function,
+                                                 revng::ranks::BasicBlock,
+                                                 revng::ranks::Instruction);
+      revng_assert(Caller.has_value());
+      Result.insert(std::get<0>(*Caller));
+    }
+  }
+}
+
+// Find callers of a given `model::DynamicFunction` via CrossRelations
+inline void insertCallers(const TupleTree<CrossRelations> &CR,
+                          std::string Entry,
+                          std::set<MetaAddress> &Result) {
+  const auto Location = location(revng::ranks::DynamicFunction, Entry);
+  const auto &Related = CR->Relations.at(Location.toString()).Related;
+  for (const yield::RelationTarget &RT : Related) {
+    if (RT.Kind == yield::RelationType::IsDynamicallyCalledFrom) {
+      auto Caller = genericLocationFromString<0>(RT.Location,
+                                                 revng::ranks::Function,
+                                                 revng::ranks::BasicBlock,
+                                                 revng::ranks::Instruction);
+      revng_assert(Caller.has_value());
+      Result.insert(std::get<0>(*Caller));
+    }
+  }
+}
+
+// Invalidate all the targets belonging to function rank
+inline void
+invalidateAllTargetsPerFunctionRank(const TupleTree<model::Binary> &Model,
+                                    std::set<MetaAddress> &Result) {
+  for (const model::Function &Function : Model->Functions)
+    Result.insert(Function.Entry);
+}
+
+namespace detail {
+
+inline void
+insertCallersAndTransitiveClosureInternal(const TupleTree<CrossRelations> &CR,
+                                          const TupleTree<model::Binary> &Model,
+                                          const std::string &Location,
+                                          std::set<MetaAddress> &Result) {
+  auto CRGraph = CR->toCallGraph();
```

Comment:
You need to hoist this so you do it only once per call to `getInvalidation`.

Reply:
Yeah, good catch, missed it.